### PR TITLE
Add simple heuristic fragment matching

### PR DESCRIPTION
**Pull request by kitten**
This is just a proposal and very ugly code

### CODE REVIEW

File: src/operations/shared.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      // A fragment is either referred to by FragmentSpread or inline
+      const def = isInlineFragment(node) ? node : ctx.fragments[getName(node)];
+
+      if (def !== undefined) {
+        const fragmentSelect = getSelectionSet(def);
+        const typeCondition = getTypeCondition(def);
+
+        const matches =
+          typeCondition === typeName ||
+          !fragmentSelect.some(node => {
```
Comment (JoviDeCroock) [inline src/operations/shared.ts]: If we quit on first occurrence we can't notify the consumer of the graphcache what exactly was wrong about the fragment. I'd prefer this for DX purposes

File: src/operations/write.ts [hunk 2]
```diff
@@ -65,8 +65,9 @@ export const write = (
   if (operation.operation === 'query') {
     writeSelection(ctx, 'Query', select, data);
   } else {
-    const isMutation = operation.operation === 'mutation';
-    writeRoot(ctx, isMutation, select, data);
+    const typename =
```
Comment (JoviDeCroock) [inline src/operations/write.ts]: Maybe we should reintroduce the helper that converts query --> Query and so on

File: src/operations/write.ts [hunk 3]
```diff
@@ -91,10 +92,11 @@ export const writeOptimistic = (
   };
 
   if (operation.operation === 'mutation') {
+    const typename = 'Mutation';
```
Comment (JoviDeCroock) [inline src/operations/write.ts]: Same here mutation --> Mutation helper

File: src/operations/shared.ts (hunk 1 above)
Comment (kitten) [reply src/operations/shared.ts]: I wouldn't actually notify though unless we have a solution for non-heuristic fragment matching. If the user uses interfaces then they'll be aware that they are, but we can't offer them any advice on when a fragment matches or doesn't match 😓

File: src/operations/write.ts (hunk 2 above)
Comment (kitten) [reply src/operations/write.ts]: probably

Comment (kitten) [reply src/operations/write.ts]: done

**Pull request merged**